### PR TITLE
[BEAM-2817] BigQuery queries are allowed to run in either BATCH or INTERACTIVE mode

### PR DESCRIPTION
**Pull request by justintumale**
Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] Make sure there is a [JIRA issue](https://issues.apache.org/jira/projects/BEAM/issues/) filed for the change (usually before you start working on it).  Trivial changes like typos do not require a JIRA issue.  Your pull request should address just this issue, without pulling in other changes.
 - [ ] Each commit in the pull request should have a meaningful subject line and body.
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue.
 - [ ] Write a pull request description that is detailed enough to understand what the pull request does, how, and why.
 - [ ] Run `mvn clean verify` to make sure basic checks pass. A more thorough check will be performed on your pull request automatically.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

---

### CODE REVIEW

**Comment by chamikaramj** [comment]
Thanks. Is this ready for review ?

**Comment by justintumale** [comment]
@chamikaramj yes, thanks.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryQuerySource.java [hunk 1]
```diff
@@ -81,12 +84,14 @@ private BigQueryQuerySource(
...
       Coder<T> coder,
-      SerializableFunction<SchemaAndRecord, T> parseFn) {
+      SerializableFunction<SchemaAndRecord, T> parseFn,
+      String priority) {
     super(stepUuid, bqServices, coder, parseFn);
     this.query = checkNotNull(query, "query");
     this.flattenResults = checkNotNull(flattenResults, "flattenResults");
     this.useLegacySql = checkNotNull(useLegacySql, "useLegacySql");
     this.dryRunJobStats = new AtomicReference<>();
+    this.priority = priority == null ? "BATCH" : priority;
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryQuerySource.java]: Let's create a new enum in BigQueryIO for BATCH and INTERACTIVE, similar to CreateDisposition.

We can also add documentation there to describe this feature.

File: sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java [hunk 2]
```diff
@@ -1898,7 +1898,8 @@ public void testBigQueryQuerySourceInitSplit() throws Exception {
         true /* useLegacySql */,
         fakeBqServices,
         TableRowJsonCoder.of(),
-        BigQueryIO.TableRowParser.INSTANCE);
+        BigQueryIO.TableRowParser.INSTANCE,
+        "BATCH");
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/test/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIOTest.java]: Please add a test that sets priority in BigQueryIO to BATCH/STREAMING and confirms that request to BQ contains the correct priority.

File: sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java [hunk 3]
```diff
@@ -534,6 +534,7 @@ public Read withTemplateCompatibility() {
       abstract Builder<T> setUseLegacySql(Boolean useLegacySql);
       abstract Builder<T> setWithTemplateCompatibility(Boolean useTemplateCompatibility);
       abstract Builder<T> setBigQueryServices(BigQueryServices bigQueryServices);
+      abstract Builder<T> setPriority(String priority);
```
Comment (chamikaramj) [inline sdks/java/io/google-cloud-platform/src/main/java/org/apache/beam/sdk/io/gcp/bigquery/BigQueryIO.java]: We should add logic to fail when an invalid priority value is set.

**Comment by jkff** [comment]
@justintumale Why was this PR closed? Seems like it was almost good to go.

**Comment by justintumale** [comment]
Hey @jkff, I did a git reset hard on my local environment and pushed the changes to my repo in order to keep in sync with the upstream, and that closed the pull request (unintentionally on my part). I was going to reopen once I was ready to push to my new local changes- but I can reopen if this has caused confusion. Thanks

**Comment by justintumale** [comment]
Update: feedback items are implemented and working on unit tests to check if the right priority is being applied. All other test cases passing

**Comment by justintumale** [comment]
Hi @chamikaramj. Ive implemented the feedback items, but it seems that even after rebasing my repo, any time I open/reopen this pull request that commits that have already been merged show up in the commit section. Any advice on how to fix this? Thanks

**Comment by jkff** [comment]
I suggest to re-create the pull request from scratch, by cherry-picking (`git cherry-pick`) the commit with the desired part of your change onto the master branch.

**Comment by chamikaramj** [comment]
Yeah, probably re-creating the PR is the easiest way out. Please follow instructions in https://beam.apache.org/contribute/contribution-guide/.

**Pull request closed without merging**